app/src/main/java/de/danoeh/antennapod/dialog/EpisodesApplyActionFragment.java

```diff
@@ -0,0 +1,414 @@
+            }
+        }
+        refreshCheckboxes();
+    }
+
+    private void checkDownloaded(boolean isDownloaded) {
+        for (FeedItem episode : episodes) {
+            if(episode.getMedia().isDownloaded() == isDownloaded) {
+                if(!checkedIds.contains(episode.getId())) {
+                    checkedIds.add(episode.getId());
+                }
+            } else {
+                if(checkedIds.contains(episode.getId())) {
+                    checkedIds.remove(episode.getId());
+                }
+            }
+        }
+        refreshCheckboxes();
+    }
+
+    private void refreshTitles() {
+        titles.clear();
+        for(FeedItem episode : episodes) {
+            titles.add(episode.getTitle());
+        }
+        mAdapter.notifyDataSetChanged();
+    }
+
+    private void refreshCheckboxes() {
+        for (int i = 0; i < episodes.size(); i++) {
+            FeedItem episode = episodes.get(i);
+            boolean checked = checkedIds.contains(episode.getId());
+            mListView.setItemChecked(i, checked);
+        }
+        ActivityCompat.invalidateOptionsMenu(EpisodesApplyActionFragment.this.getActivity());
+    }
+
+    private void queueChecked() {
+        LongList orderedIds = new LongList();
+        for(FeedItem episode : episodes) {
+            if(checkedIds.contains(episode.getId())) {
+               orderedIds.add((episode.getId()));
+            }
+        }
+        DBWriter.addQueueItem(getActivity(), false, orderedIds.toArray());
+        close();
+    }
+
+    private void markedCheckedPlayed() {
+        DBWriter.markItemRead(getActivity(), true, checkedIds.toArray());
+        close();
+    }
+
+    private void markedCheckedUnplayed() {
+        DBWriter.markItemRead(getActivity(), false, checkedIds.toArray());
+        close();
+    }
+
+    private void downloadChecked() {
+        // download the check episodes in the same order as they are currently displayed
+        List<FeedItem> toDownload = new ArrayList<FeedItem>(checkedIds.size());
+        for(FeedItem episode : episodes) {
+            if(checkedIds.contains(episode.getId())) {
+                toDownload.add(episode);
+            }
+        }
+        try {
+            DBTasks.downloadFeedItems(getActivity(), toDownload.toArray(new FeedItem[0]));
+        } catch (DownloadRequestException e) {
+            e.printStackTrace();
+            DownloadRequestErrorDialogCreator.newRequestErrorDialog(getActivity(), e.getMessage());
+        }
+        close();
+    }
+
+    private void deleteChecked() {
+        for(long id : checkedIds.toArray()) {
+            FeedItem episode = findById(id);
+            DBWriter.deleteFeedMediaOfItem(getActivity(), episode.getMedia().getId());
+        }
+        close();
+    }
+
+    private void close() {
+        getActivity().getSupportFragmentManager().popBackStack();
+    }
+
+    private FeedItem findById(long id) {
+        for(int i=0; i < episodes.size(); i++) {
```

Comment:
Not that it matters, but why no love for Java's foreach?

for (FeedItem episode : episodes) {
  if (episode.getId() == id) {
    return episode;
  }
}

Alternatively, should we be using a hashtable instead?  Some feeds are hundreds of episodes.  (Though I could see it not mattering in practice).